### PR TITLE
Webhook: ignore deploying some environments

### PR DESCRIPTION
**Pull request by nike38rus**
It's roughly like https://github.com/puppetlabs/r10k/issues/237 but on the webhook side and since that issue is not yet resolved for 2+ years I guess it will be good to implement it in this wonderful module ;)

I tested it in my environment with Puppet 4.8.2 and seems that it's working well

### CODE REVIEW

File: manifests/webhook/config.pp [hunk 1]
```diff
@@ -43,6 +43,7 @@
   $manage_symlink             = false,
   $configfile_symlink         = '/etc/webhook.yaml',
   $enable_mutex_lock          = $r10k::params::webhook_enable_mutex_lock,
+  Array $ignore_environments  = $r10k::params::webhook_ignore_environments,
```
Comment (bastelfreak) [inline manifests/webhook/config.pp]: yeeey Datatypes \o/

**Comment by bastelfreak** [comment]
Hey @nike38rus, thanks for the PR. @rnelson0 can you take a look as well?

File: manifests/webhook/config.pp (hunk 1 above)
Comment (oranenj) [inline manifests/webhook/config.pp]: Use Array[String] instead?

File: templates/webhook.bin.erb [hunk 2]
```diff
@@ -173,6 +179,24 @@ end
...
 
+    # Ignore environments that we don't care about e.g. feature or bugfix branches
+    def ignore_env?(env)
+      list = $config['ignore_environments']
+      return false if list.nil? or list.empty?
+
+      list.each do |l|
+        # Even unquoted array elements wrapped by slashes becomes strings after YAML parsing
+        # So we need to convert it into Regexp manually
+        if l =~ /\/.+\//
```
Comment (oranenj) [inline templates/webhook.bin.erb]: shouldn't this have start/end anchors to prevent matching things like `foo/bar/`?

Comment (nike38rus) [inline templates/webhook.bin.erb]: Good catch, thanks!

**Comment by oranenj** [comment]
Tests seem to be failing due to a lint complaint unrelated to this change.

File: README.md [hunk 3]
```diff
@@ -533,6 +533,22 @@ class { '::r10k::webhook':
 }
 ```
 
+### Ignore deploying some environments
+
+If you need to configure webhook to not trigger r10k when changes pushed in some branch or repository, you can list them in
+`r10k::webhook::config::ignore_environments` parameter as array.
+There is ability to specify it as a regular expression by enclosing it in forward slashes.
```
Comment (rnelson0) [inline README.md]: `There is an ability`

File: README.md [hunk 4]
```diff
@@ -533,6 +533,22 @@ class { '::r10k::webhook':
...
+
+Here is example showing that test branch in dev repository and all feature branches in all repositories will be skipped:
```
Comment (rnelson0) [inline README.md]: `Here is an example where the test branch...`

Comment (rnelson0) [inline README.md]: `and all branches including the word 'feature'` - is that correct?

File: templates/webhook.bin.erb [hunk 5]
```diff
@@ -160,11 +160,17 @@ end
...
     if prefix.nil? or prefix.empty? or branch.nil? or branch.empty?
-      deploy(normalize(branch))
+      env = normalize(branch)
     else
-      deploy(normalize("#{prefix}_#{branch}"))
+      env = normalize("#{prefix}_#{branch}")
     end
 
+    if ignore_env?(env)
+      $logger.info("Ignoring #{env} environment deploying")
```
Comment (rnelson0) [inline templates/webhook.bin.erb]: How about 'Skipping deployment of environment #{env} according to ignore_environments configuration' or similar?

**Comment by nike38rus** [comment]
@rnelson0 done!

**Comment by rnelson0** [comment]
@nike38rus the webhook.yaml content test willl need updated, see https://travis-ci.org/voxpupuli/puppet-r10k/jobs/219142272#L3093 for what has changed. Sorry I missed that previously!

**Comment by nike38rus** [comment]
@rnelson0 done, all green :)

**Comment by rnelson0** [comment]
@nike38rus Thank you for this fix!
Are you the same Nikita who submitted #354, and if so, does this address your concerns there so we can close it out?

**Comment by nike38rus** [comment]
@rnelson0 there's no big deal ;)
Unfortunately (or not :) ) I'm not the same Nikita as @luckyraul so I can't help with this issue

**Comment by rnelson0** [comment]
Ha, it was coincidental timing I guess. Thanks again!

**Comment by luckyraul** [comment]
different problems =)

**Pull request merged**